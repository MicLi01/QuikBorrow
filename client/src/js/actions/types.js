export const GET_ERRORS = "GET_ERRORS";
export const USER_LOADING = "USER_LOADING";
export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const SEARCH = "SEARCH";
export const GET_USER_PROFILE = "GET_USER_PROFILE";

// Products
export const FETCH_PRODUCTS = "fetch_products";
export const FETCH_PRODUCT = "fetch_product";
export const BORROW_PRODUCT_FETCH = 'BORROW_PRODUCT_FETCH';
export const BORROW_PRODUCT_FINISHED = 'BORROW_PRODUCT_FINISHED';

