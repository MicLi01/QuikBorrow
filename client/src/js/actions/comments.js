import axios from "axios";
import { CREATE_COMMENT, GET_COMMENTS } from './types';

/**
 * @typedef {import('redux').Dispatch} DispatchFunction
 */

/**
 * Create a comment
 * 
 * @param {Object} comment - The Comment object to create
 * @returns {DispatchFunction}
 */
export const createComment = (comment) => dispatch => {
  return axios
    .post("/api/comments", comment)
    .then(res => {
    	console.log(res);
    	console.log(comment);
    })
    .catch(err => {
    	console.log(err);
  		dispatch({
  			type: CREATE_COMMENT.ERROR,
  			payload: err.response.data
  		})
    });
};

/**
 *  Get comments associated with product
 *  @param {String} id - object id of product
 *  @returns {DispatchFunction} dispatch - used to dispatch actions
 */
export const getComments = (id) => dispatch => {
  return axios
    .get(`/api/comments/${id}`)
    .then(res => {
      dispatch({
        type: GET_COMMENTS.FINISHED,
        payload: res.data
      });
    })
    .catch(err => {
      dispatch({
        type: GET_COMMENTS.ERROR,
        payload: err.response.data
      })
    });
};