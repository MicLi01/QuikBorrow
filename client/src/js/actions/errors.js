import { CLEAR_ALL_ERRORS } from './types'

/**
 * Clears all error data in error state for all error types
 */
 export const clearAllErrors = () => {
     return { type: CLEAR_ALL_ERRORS }
 }
