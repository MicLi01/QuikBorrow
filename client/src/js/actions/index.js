import axios from "axios";
import setAuthToken from "../utils/setAuthToken";
import jwt_decode from "jwt-decode";
import { FETCH_PRODUCTS, FETCH_PRODUCT, SET_CURRENT_USER, USER_LOADING, SEARCH_PRODUCTS, GET_USER, REGISTER_USER, LOGIN_USER } from "./types";
export * from './products';

// Fetch all products
export const fetchProducts = () => {
  return async function(dispatch) {
    const res = await axios.get("/api/items");
    dispatch({ type: FETCH_PRODUCTS, payload: res.data });
  };
};

// Get one product
export const fetchProduct = itemId => {
  return async function(dispatch) {
    const res = await axios.get(`/api/items/${itemId}`);
    dispatch({ type: FETCH_PRODUCT, payload: res.data });
  };
};

/**
 *  Search by query for items in database by calling {@link Action} to backend
 *  @param {String} query - query string to search
 *  @returns {Function} dispatch - used to dispatch actions
 */
export const searchProducts = query => 
  /**
   *  @param {Function} dispatch - used to dispatch actions
   *  @returns {Promise<Object>} - response from an {@link Action}
   */
  dispatch => {
    return axios
        .get("/api/items/search", { params: { param: query} })
        .then(res => {
          dispatch({ type: SEARCH_PRODUCTS.FETCH, payload: res.data });
        })
        .catch(err => {
          dispatch({
            type: SEARCH_PRODUCTS.ERROR,
            payload: err.response.data
          })
        });
  };

// Register User
export const registerUser = (userData, history) => dispatch => {
  return axios
    .post("/api/users/register", userData)
    .then(res => history.push("/login")) // re-direct to login on successful register
    .catch(err =>
      dispatch({
        type: REGISTER_USER.ERROR,
        payload: err.response.data
      })
    );
};

// Login - get user token
export const loginUser = userData => dispatch => {
  return axios
    .post("/api/users/login", userData)
    .then(res => {
      // Save to localStorage
      // Set token to localStorage
      const { token } = res.data;
      localStorage.setItem("jwtToken", token);
      // Set token to Auth header
      setAuthToken(token);
      // Decode token to get user data
      const decoded = jwt_decode(token);
      // Set current user
      dispatch(setCurrentUser(decoded));
    })
    .catch(err =>
      dispatch({
        type: LOGIN_USER.ERROR,
        payload: err.response.data
      })
    );
};

// Set logged in user
export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

// User loading
export const setUserLoading = () => {
  return {
    type: USER_LOADING
  };
};

// Log user out
export const logoutUser = () => dispatch => {
  // Remove token from local storage
  localStorage.removeItem("jwtToken");
  // Remove auth header for future requests
  setAuthToken(false);
  // Set current user to empty object {} which will set isAuthenticated to false
  dispatch(setCurrentUser({}));
};

export const getUserProfile = userId => {
  return async function(dispatch) {
    try {
    const res = await axios.get(`/api/users/${userId}`);
    dispatch({ type: GET_USER.FINISHED, payload: res.data });
  }
  catch(error) {
    console.log(error)
    dispatch({
      type: GET_USER.ERROR, payload: error
      });
    }
  }
};
