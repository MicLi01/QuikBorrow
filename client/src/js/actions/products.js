import axios from "axios";
import {
    FETCH_PRODUCTS,
    FETCH_PRODUCT,
    SEARCH_PRODUCTS,
    REQUEST_BORROW_PRODUCT,
    FETCH_TRANSACTIONS
} from './types'

/**
 * @typedef {import('redux').Dispatch} DispatchFunction
 */

/**
 * Borrow an item via fetch
 * 
 * @param {Object} item - The Item object to borrow
 * @param {string} msg - The message for the lender
 * @returns {DispatchFunction}
 */
export const requestBorrowProductFetch = (item, msg) => async (dispatch, getState) => {
    const state = getState()
    const body = {
        borrowerId: state.auth.user.id,
        lenderId: item.user,
        itemId: item._id,
        msg
    }
    return axios.post(`/api/transactions`, body)
        .then(({ data }) => {
            dispatch({ type: REQUEST_BORROW_PRODUCT.FINISHED, payload: data })
        })
        .catch(err => {
            dispatch({ type: REQUEST_BORROW_PRODUCT.ERROR, payload: err.response.data })
        })
}

/**
 * Fetch all products
 * @returns {DispatchFunction}
 */
export const fetchProducts = () => {
  return async function(dispatch) {
    const res = await axios.get("/api/items");
    dispatch({ type: FETCH_PRODUCTS, payload: res.data });
  };
};

/**
 * Get one product
 * 
 * @param {Object} itemId - The id of the item to retrieve
 * @returns {DispatchFunction}
 */
export const fetchProduct = itemId => {
  return async function(dispatch) {
    const res = await axios.get(`/api/items/${itemId}`);
    dispatch({ type: FETCH_PRODUCT, payload: res.data });
  };
};

/**
 *  Search by query for items in database by calling {@link Action} to backend
 *  @param {String} query - query string to search
 *  @returns {Function} dispatch - used to dispatch actions
 */
export const searchProducts = query => 
  /**
   *  @param {Function} dispatch - used to dispatch actions
   *  @returns {Promise<Object>} - response from an {@link Action}
   */
  dispatch => {
    return axios
        .get("/api/items/search", { params: { param: query} })
        .then(res => {
          dispatch({ type: SEARCH_PRODUCTS.FINISHED, payload: res.data });
        })
        .catch(err => {
          dispatch({
            type: SEARCH_PRODUCTS.ERROR,
            payload: err.response.data
          })
        });
  };

  /**
 * Get all current user transactions
 * 
 * @param {Object} id - The user id
 * @returns {Function} dispatch - used to dispatch actions
 */

 export const fetchTransactions = (id) => 
  /**
   *  @param {Function} dispatch - used to dispatch actions
   *  @returns {Promise<Object>} - response from an {@link Action}
   */
  dispatch => {
    return axios.post('/api/transactions/getTransactions', {userId : id})
        .then(res => dispatch({ type: FETCH_TRANSACTIONS.FINISHED, payload: res.data }))
        .catch(err => dispatch({ type: FETCH_TRANSACTIONS.ERROR, payload: err.response.data }))
 }

