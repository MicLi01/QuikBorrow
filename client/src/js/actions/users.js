import axios from "axios";
import setAuthToken from "../utils/setAuthToken";
import jwt_decode from "jwt-decode";
import { SET_CURRENT_USER, USER_LOADING, REGISTER_USER, LOGIN_USER, GET_USER } from "./types";

/**
 * @typedef {import('redux').Dispatch} DispatchFunction
 */

/**
 *	Register user 
 *	@param {Object} userData - User credentials to create an account
 *	@param {Object} history - used to navigate to different pages
 *	@returns {DispatchFunction}
 */
export const registerUser = (userData, history) => dispatch => {
  return axios
    .post("/api/users/register", userData)
    .then(res => history.push("/login")) // re-direct to login on successful register
    .catch(err =>
      dispatch({
        type: REGISTER_USER.ERROR,
        payload: err.response.data
      })
    );
};

/**
 *	Login user 
 *	@param {Object} userData - User credentials to login
 *	@returns {DispatchFunction}
 */
export const loginUser = userData => dispatch => {
  return axios
    .post("/api/users/login", userData)
    .then(res => {
      const { token } = res.data;
      localStorage.setItem("jwtToken", token);
      // Set token to Auth header
      setAuthToken(token);
      // Decode token to get user data
      const decoded = jwt_decode(token);
      // Set current user
      dispatch(setCurrentUser(decoded));
    })
    .catch(err =>
      dispatch({
        type: LOGIN_USER.ERROR,
        payload: err.response.data
      })
    );
};

/**
 *	Set current user
 *	@param {Object} decoded - decoded jwtToken containing user info
 *	@returns {Object}
 */
export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

/**
 *	Set user loading
 *	@returns {Object}
 */
export const setUserLoading = () => {
  return {
    type: USER_LOADING
  };
};

/**
 *	Logout user by removing auth token and saved info in storage
 *	@returns {DispatchFunction}
 */
export const logoutUser = () => dispatch => {
  // Remove token from local storage
  localStorage.removeItem("jwtToken");
  // Remove auth header for future requests
  setAuthToken(false);
  // Set current user to empty object {} which will set isAuthenticated to false
  dispatch(setCurrentUser({}));
};

/**
 *	Get user profile
 *	@param {Object} userId - get profile based on userId
 *	@returns {DispatchFunction}
 */
export const getUserProfile = userId => {
  return async function(dispatch) {
    try {
    const res = await axios.get(`/api/users/${userId}`);
    dispatch({ type: GET_USER.FINISHED, payload: res.data });
  }
  catch(error) {
    console.log(error)
    dispatch({
      type: GET_USER.ERROR, payload: error
      });
    }
  }
};