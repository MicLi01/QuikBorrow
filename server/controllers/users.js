const userServices = require('../services/users');

// eslint-disable-next-line valid-jsdoc
/** @type {import('express').RequestHandler} */
const login = (req, res, next) => {
  const payload = {
    id: req.user.id,
    name: req.user.name,
  };

  return userServices.getJwtToken(payload)
      .then((token) => res.json({success: true, token: 'Bearer ' + token}))
      .catch(next);
};

// eslint-disable-next-line valid-jsdoc
/** @type {import('express').RequestHandler} */
const register = async (req, res, next) => {
  const {name, email, password} = req.body;
  try {
    const hashedPassword = await userServices.generateHash(password);
    const user = {name, email, password: hashedPassword};
    const createdUser = await userServices.createUser(user);
    res.json(createdUser);
  } catch (err) {
    next(err);
  }
};

const get = (req, res, next) => {
  return userServices.findUser(req.params.userId)
      .then((user) => res.json(user))
      .catch(next);
};

module.exports = {
  login,
  register,
  get,
};
