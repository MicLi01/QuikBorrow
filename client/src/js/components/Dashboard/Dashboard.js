import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { logoutUser } from "../../actions/users";

export class Dashboard extends Component {
  componentDidMount () {
    if (this.props.auth.isAuthenticated !== true) {
      this.props.history.push('/login')
    }
  }
  componentWillReceiveProps(nextProps) {
      if (nextProps.auth.isAuthenticated !== true) {
          this.props.history.push("/login");
      }
  }
  onLogoutClick = e => {
    e.preventDefault();
    this.props.logoutUser();
  };
    render() {
    const { user } = this.props.auth;
    return (
      <div>
        <h4>
          <b>Welcome</b>. You are logged in.
        </h4>
        <button onClick={this.onLogoutClick}>Logout</button>
      </div>
    );
  }
}

Dashboard.propTypes = {
  logoutUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { logoutUser }
)(Dashboard);