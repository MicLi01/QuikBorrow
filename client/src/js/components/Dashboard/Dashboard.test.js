import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Dashboard } from './Dashboard';
import { logoutUser } from "../../actions/users";

Enzyme.configure({ adapter: new Adapter() });
jest.mock('../../actions/users');

describe('Dashboard', () => {
	const params = {
		logoutUser : logoutUser,
		auth: {
			isAuthenticated: true,
			user: {},
			loading: false,
		}
	}

	test('renders', () => {
		const component = renderer.create(
			<MemoryRouter>
				<Dashboard logoutUser = {params.logoutUser} auth = {params.auth} />
			</MemoryRouter>
		).toJSON();
		
		expect(component).toMatchSnapshot();
	});

	test('redirect to login', () => {
		const notAuth = {
			isAuthenticated: false,
			user: {},
			loading: false,
		}
		const historyMock = { push: jest.fn() };
		let wrapper = shallow(<Dashboard logoutUser = {params.logoutUser} auth = { notAuth } history = {historyMock} />);
		wrapper.setProps({ isAuthenticated: false });
		expect(historyMock.push.mock.calls[0]).toEqual(['/login']);
	});

	test('logout', () => {
		let wrapper = shallow(<Dashboard logoutUser = {params.logoutUser} auth = {params.auth} />);
		wrapper.find('button').simulate('click', {preventDefault() {}});
		expect(wrapper.instance().props.logoutUser).toHaveBeenCalled();
	});
});