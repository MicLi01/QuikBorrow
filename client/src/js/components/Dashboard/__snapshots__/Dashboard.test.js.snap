// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders 1`] = `
<div>
  <h4>
    <b>
      Welcome
    </b>
    . You are logged in.
  </h4>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</div>
`;
